### PR TITLE
Implement a material search

### LINKED ISSUE #56: semantic search for material
My plan is to use openai to make vectors, save them to a local qdrant vector db and search terms are also vectorised and compared to the db.

### PR DESCRIPTION
**Pull request by alasdairwilson**
Vector embedding based search for the material directory, powered by openai:
search_demo

So this is done by first parsing the markdown to remove some directives and then split the files into sections, take the text from the sections and vectorise them with openai, save this into a QDRANT database (hosted on fly.io or otherwise).

When the user searches we first make an api call to open ai to return a vector embedding and then search our material qdrant database for the closest results, these are returned to the frontend.

There is a secondary thing going on here, the maintaining of up-to-date material and up-to-date vector DB. So this also uses a cron job to, at midnight, update the MATERIAL_DIR with the main branch of the github material repo, check if the material has changed since last the vectors were updated and, if so, re-vectorise and replace the contents of the qdrant database.

closes #56 #8 

There are some obvious follow-ons:

- Add anchor tags to the markdown pages ( remark-slug/remark-toc/remark-autolink-headings? or see https://github.com/remarkjs/react-markdown/issues/404 )
- Change the search to use remark-rehype to parse the files.
- Move the qdrant docker file and fly.toml to a separate repo.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (martinjrobins) [inline .vscode/launch.json:1]: I don't think you meant to commit this file.

File: Dockerfile [hunk 2]
```diff
@@ -1,9 +1,10 @@
 # Install dependencies only when needed
-FROM node:16-alpine AS builder
+FROM node:20-alpine AS builder
 # Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
 RUN apk add --no-cache libc6-compat
 WORKDIR /app
 COPY . .
+RUN chmod +x entrypoint.sh
```
Comment (martinjrobins) [inline Dockerfile:7]: not sure why this is neccessary to do in the dockerfile? is not entrypoint.sh already executable?

File: lib/actions/searchMaterial.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+// GET /api/search
+const searchQuery = async (query: string) => {
+  const apiPath = `${basePath}/api/search/`
+  const requestOptions = {
+      method: 'POST',
+      headers: { 'Content-Type': 'text/html' },
+      body: query,
+  };
+  console.log(query)
```
Comment (martinjrobins) [inline lib/actions/searchMaterial.ts:11]: can we remove these logs, I assume they were for debugging?

File: pages/api/search.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import type { NextApiRequest, NextApiResponse } from 'next'
+import { getEmbedding } from 'lib/search/createVectors'
+import { searchVector, SearchResult } from 'lib/search/vectorDb';
+
+ 
+export default function searchHandler(
+        req: NextApiRequest,
+        res: NextApiResponse<SearchResult[]>
+    ) {
+        console.log('request received', req.body)
```
Comment (martinjrobins) [inline pages/api/search.ts:10]: probably get rid of this as well?

**Review by martinjrobins — DISMISSED** [review]
this looks great, thanks @alasdairwilson . Just a few minor things below then can be merged once you are happy.

File: Dockerfile (hunk 2 above)
Comment (alasdairwilson) [reply Dockerfile:7]: It had been my understanding that docker stripped the execution bit when you copied and so I was in the habit of chmod +x to reinstate it but it turns out this is if you build in windows.

**Comment by alasdairwilson** [comment]
ready to merge, just needs an approval ( the changes dismissed your prior review )

**Review by martinjrobins — APPROVED** [review]
all happy with the changes, thanks @alasdairwilson

**Pull request merged**